### PR TITLE
ci: more parallel tests

### PR DESCRIPTION
**Pull request by georglauterbach**
# Description

More parallel tests -> reduce time it takes to run CI. Profit.

## Type of change

- [x] Improvement (non-breaking change that does improve existing functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [ ] If necessary I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by georglauterbach** [comment]
@polarathene I had hoped you could help me here: in the test `Traefik 'acme.json' (*.example.test)"` in the file `test/tests/parallel/set2/ssl_tls/ssl_letsencrypt.bats`, I cannot figure out why the test is failing. Could you have a look? This is what I see locally:

```console
$ make clean generate-accounts tests/parallel/set2
ssl_letsencrypt.bats
 ✓ SSL Let'sEncrypt: Should default to HOSTNAME (mail.example.test) in 14268ms [14268]
 ✓ SSL Let'sEncrypt: Should fallback to DOMAINNAME (example.test) in 14405ms [14405]
 ✗ SSL Let'sEncrypt: Traefik 'acme.json' (*.example.test) in 30875ms [30875]
   (from function `assert_output' in file test/test_helper/bats-assert/src/assert_output.bash, line 186,
    from function `_negotiate_tls' in file test/helper/tls.bash, line 48,
    from function `_should_succesfully_negotiate_tls' in file test/helper/tls.bash, line 26,
    from function `_acme_wildcard' in file test/tests/parallel/set2/ssl_tls/ssl_letsencrypt.bats, line 167,
    in test file test/tests/parallel/set2/ssl_tls/ssl_letsencrypt.bats, line 178)
     `_acme_wildcard' failed
   [   INF   ]  mail.example.test is up and running
   [  DEBUG  ]  2022-12-19 23:46:15  Changedetector is ready
   
   -- output does not contain substring --
   substring (1 lines):
     Verification: OK
   output (49 lines):
     CONNECTED(00000003)
     ---
     Certificate chain
      0 s:CN = Smallstep Leaf
        i:CN = Smallstep Root CA
     ---
     Server certificate
     -----BEGIN CERTIFICATE-----
     MIIDTjCCAjagAwIBAgIRAMjWRj4nHDOUhio1Pc+1Om8wDQYJKoZIhvcNAQELBQAw
     HDEaMBgGA1UEAxMRU21hbGxzdGVwIFJvb3QgQ0EwHhcNMjEwMTAxMDAwMDAwWhcN
     MzEwMTAxMDAwMDAwWjAZMRcwFQYDVQQDEw5TbWFsbHN0ZXAgTGVhZjCCASIwDQYJ
     KoZIhvcNAQEBBQADggEPADCCAQoCggEBALQA64PCkaIQLtzNOL3RcUwFT23canZJ
     eQm4Ai6UKPE/CoHlCyc4PvLmstdegD6p7wzNX7Wp9wlrGhl0mW6fC9W5UU7ziSA9
     ffqrAjUzW0MGPusqFERq4XUY6vMIJ/DkoNxNDcsbRRcF7e3x/q60vLhnjwSD7nw6
     s1q1A08MPaz3GOC+etxb83rCa+SIVhHv6t5ScZR/IccINC5BDnQRl9ED96jaRcuE
     DrX59/KUyIicOCYy6lYvqEdxxb9dUPgOYLvUiicgYWhHja56uF23nhagTz+eYMI4
     vZPIeGo2BkvT0rcyBJ7QnGOs34UvoX7bK4Xu9re2fgU+dKvVAdnjZ9UCAwEAAaOB
     jTCBijAOBgNVHQ8BAf8EBAMCBaAwHQYDVR0lBBYwFAYIKwYBBQUHAwEGCCsGAQUF
     BwMCMB0GA1UdDgQWBBR4Ut8IBrlcCqeanmb3/eip4J4dwjAfBgNVHSMEGDAWgBQd
     z+3MFEz93EC48t0r750mEC8/wzAZBgNVHREEEjAQgg4qLmV4YW1wbGUudGVzdDAN
     BgkqhkiG9w0BAQsFAAOCAQEAl8+2+jByETsTMJXBc0U4DHdUqEF+nffF5ckMu6zx
     DRMhC8IuXicQ9cd4dlonl2joaP0Y1oOvhZE3JBtVAKdKjDIuiuJCMhsJMvCFimX7
     pCaErMVtIYOrobG4mYo0rQv7Faq7X8xE8/m76yaGQedx6jaeQEDqI72M042UIa8R
     QrexAGIYcVBGSmOoKBXLbgTgY+kw1jweWM9ql9DqjKymdoO70SWpefJwc1k5FAR1
     Xi1eu2RNOAsGZniWsIWitPmInhOkAq3vaJ/clepL0+TUA6Kl1qIh7z2sVK64ms5E
     HKF69RdtrA1UH0gHlQkAueTW8LI2EUYBuZZ+gEMSVczdbQ==
     -----END CERTIFICATE-----
     subject=CN = Smallstep Leaf
   
     issuer=CN = Smallstep Root CA
   
     ---
     No client certificate CA names sent
     Peer signing digest: SHA256
     Peer signature type: RSA-PSS
     Server Temp Key: X25519, 253 bits
     ---
     SSL handshake has read 1653 bytes and written 396 bytes
     Verification error: unable to verify the first certificate
     ---
     New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
     Server public key is 2048 bit
     Secure Renegotiation IS NOT supported
     Compression: NONE
     Expansion: NONE
     No ALPN negotiated
     Early data was not sent
     Verify return code: 21 (unable to verify the first certificate)
     ---
   --
   
   dms-test-ssl_letsencrypt-3
ssl_manual.bats
 ✓ SSL (manual): ENV vars provided are valid files in 339ms [339]
 ✓ SSL (manual): manual configuration is correct in 475ms [475]
 ✓ SSL (manual): manual configuration copied files correctly  in 379ms [379]
 ✓ SSL (manual): manual cert works correctly in 385ms [385]
 ✓ SSL (manual): manual cert changes are picked up by check-for-changes in 3092ms [3092]
/home/georg/documents/git/hub/docker-mailserver/test/tests/parallel/set2/template.bats
 ✓ template: default check in 131ms [131]

9 tests, 1 failure in 48 seconds

make: *** [Makefile:55: tests/parallel/set2] Error 1
```

**Comment by polarathene** [comment]
> ```
> Verify return code: 21 (unable to verify the first certificate)
> ```

The test output is getting this error instead. So something you've changed has has introduced that failure.

Looking at where the test fails and prior to that, it seems like all the files were checked and verified as being there, so this is a bit odd :thinking: 

```bash
  # Test that `acme.json` also works with wildcard certificates:
  # Additionally tests that SSL_DOMAIN is prioritized when `letsencrypt/live/` already has a HOSTNAME dir available.
  # Wildcard `*.example.test` should extract to `example.test/` in `letsencrypt/live/`:
  function _acme_wildcard() {
    _should_extract_on_changes 'example.test' "${LOCAL_BASE_PATH}/wildcard/rsa.acme.json"
    _should_have_service_restart_count '2'
    # As the FQDN has changed since startup, the Postfix + Dovecot configs should be updated:
    _should_have_valid_config 'example.test' 'key.pem' 'fullchain.pem'
    local WILDCARD_KEY_PATH="${LOCAL_BASE_PATH}/wildcard/key.rsa.pem"
    local WILDCARD_CERT_PATH="${LOCAL_BASE_PATH}/wildcard/cert.rsa.pem"
    _should_have_expected_files 'example.test' "${WILDCARD_KEY_PATH}" "${WILDCARD_CERT_PATH}"
    # These two tests will confirm wildcard support is working, the supported SANs changed:
    # Before (_acme_rsa cert):      `DNS:example.test, DNS:mail.example.test`
    # After  (_acme_wildcard cert): `DNS:*.example.test`
    # The difference in support is:
    # - `example.test` should no longer be valid.
    # - `mail.example.test` should remain valid, but also allow any other subdomain/hostname.
    _should_succesfully_negotiate_tls 'mail.example.test'
```

I guess I would need to find some time to run the test locally and reproduce the failure and further inspect where that's going wrong.

Did you want to stash / extract that test update for a separate PR and get everything else reviewed and approved first?

**Comment by georglauterbach** [comment]
> Did you want to stash / extract that test update for a separate PR and get everything else reviewed and approved first?

Probably a good idea; will tend to it.

**Comment by georglauterbach** [comment]
Should be ready for review now.

**Comment by polarathene** [comment]
> I cannot figure out why the test is failing. Could you have a look?

I investigated this.

When you changed from the `setup_file()` that initialized setup for all test cases, to a method each one would call at different parts within their test cases, you introduced a bug as those initial vars were now defined after overrides in the test case had been declared (_specifically `TEST_CA_CERT` above `_prepare()` in the traefik `acme.json` test case_).

It seems you would now need to move that line into `_prepare()` after you call the `run_setup_for_individual_test()` method you created, and then use `export` instead of `local`?

That seems to fix the issue :)

**Comment by georglauterbach** [comment]
> > I cannot figure out why the test is failing. Could you have a look?
>
> I investigated this.

Nice!

> Seems you would now need to move that line into `_prepare()` after you call the `run_setup_for_individual_test()` method you created, and then use `export` instead of `local`?

Which line exactly? Do you have a fixed version of the file at hand you can just share? This way there's no uncertainty :)

**Comment by polarathene** [comment]
> Which line exactly?

There's a single `TEST_CA_CERT` line in that file, and a single `_prepare()` function too :sweat_smile: 

I can take care of it after I've debugged what's going wrong with the `reload` PR changes.

File: test/tests/parallel/set1/spam_virus/spam_junk_folder.bats
Comment (polarathene) [inline test/tests/parallel/set1/spam_virus/spam_junk_folder.bats:41]: Same concern here.

Note that the change from hard-coded distinct container names to now refer to `${CONTAINER_NAME}` (_and not an explicit variant defined at the top_) in test cases, now makes them a bit ambiguous? (_**EDIT:** Non-issue at present_)

~~This places more importance in keeping the test files segmented due to declaration order, but I guess in future, these can be technically split into separate test files beside each other?~~ **Nevermind**, I just noticed this test is only two separate test cases :sweat_smile:

File: test/tests/parallel/set1/spam_virus/spam_junk_folder.bats
Comment (polarathene) [inline test/tests/parallel/set1/spam_virus/spam_junk_folder.bats:54]: Is there a reason for defining this within the test case instead of at the top to apply to all?

File: test/tests/parallel/set1/spam_virus/undef_spam_subject.bats
Comment (polarathene) [inline test/tests/parallel/set1/spam_virus/undef_spam_subject.bats:34]: Required to resolve conflict from recent PR merge. This ENV wasn't doing anything to support the test.

```suggestion
```

File: test/tests/parallel/set1/spam_virus/fail2ban.bats
Comment (polarathene) [inline test/tests/parallel/set1/spam_virus/fail2ban.bats:94]:
```suggestion
  # need a bit more sleep as the test below that (which restarts F2B) may interfere
  sleep 30
```

Which test is being referred to here? Ideally we can avoid a 30 second sleep..

File: test/tests/parallel/set1/spam_virus/fail2ban.bats
Comment (polarathene) [inline test/tests/parallel/set1/spam_virus/fail2ban.bats:105]: Are the tests cases within the same file running at the same time?

My recollection was that multiple `.bats` files were being run in parallel on a system, not individual test cases.

**Review by polarathene — CHANGES_REQUESTED** [review]
Review is mostly completed. I'll follow up with the letsencrypt test changes within an hour as I have to dash out for a bit.

Great work so far!

---

In the commit history, one diff for `disabled_clamav_spamassassin.bats` had this comment:

```bash
# TODO: find a better way to know when we have waited long enough
# for ClamAV to should have come up, if it were enabled
```

Was this relevant to the `clamav.bats` test at all and it's current approach for this?

File: test/tests/parallel/set1/spam_virus/undef_spam_subject.bats
Comment (polarathene) [reply test/tests/parallel/set1/spam_virus/undef_spam_subject.bats]: I've resolved this.

**Comment by polarathene** [comment]
I thought the commit for `SASL_PASSWD` might have resolved the conflict issue :sweat_smile: Perhaps I need to rewrite history to tackle that. Don't worry about it if you're ok with me rewriting the history afterwards :)

Due to all the tests being renamed/moved, and Github not being able to properly communicate that kind of diff, I'd like to **avoid merge commits** and once the PR is approved for merging:

- Rebase
- Rewrite the history (_which seems to be tricky when merge commits from master are involved? :man_shrugging: )
- Then a merge commit into master instead of the usual squash.

File: test/tests/parallel/set1/spam_virus/clamav.bats
Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/clamav.bats:29]: Not really, it's just more consistent with the rest of the test cases I looked at.

File: test/tests/parallel/set1/spam_virus/spam_junk_folder.bats
Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/spam_junk_folder.bats:41]: > Same concern here.
> 
> Note that the change from hard-coded distinct container names to now refer to `${CONTAINER_NAME}` (_and not an explicit variant defined at the top_) in test cases, now makes them a bit ambiguous? (_**EDIT:** Non-issue at present_)

Agreed, it is _a bit_ more ambiguous.

> [...] **Nevermind**, I just noticed this test is only two separate test cases 😅

So is this resolved?

File: test/tests/parallel/set1/spam_virus/spam_junk_folder.bats
Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/spam_junk_folder.bats:54]: Probably not, will have a look.

File: test/tests/parallel/set1/spam_virus/fail2ban.bats
Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/fail2ban.bats:94]: The current issue I faced with porting the F2B tests is that the test that restarts F2B can interfere with other tests apparently, when they are executed at the sime time. I was under the impression BATS would run two different test _files_ at the same time, but from what I was seeing yesterday, `@test`s were run in parallel too.

Uhhh, ohhh; I just wrote the above text, searched for it and found the solution: `--no-parallelize-within-files`. I will provide an update for this.

File: test/tests/parallel/set1/spam_virus/fail2ban.bats
Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/fail2ban.bats:105]: Haha, see my answer above: <https://github.com/docker-mailserver/docker-mailserver/pull/2938#discussion_r1056239775> :D

**Comment by georglauterbach** [comment]
Alright @polarathene, I will go ahead and apply some feedback; you can then take over to apply the fixes to the SSL tests and rewrite history :)

File: test/tests/parallel/set1/spam_virus/disabled_clamav_spamassassin.bats
Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/disabled_clamav_spamassassin.bats:47]: Some of this feedback is about the test itself; I would like to not touch the test if not require in order to not bloat the diff. We can take care of the `ps aux --forest | grep -v grep | grep '/usr/sbin/clamd'` stuff in a later PR :)

File: test/tests/parallel/set1/spam_virus/fail2ban.bats
Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/fail2ban.bats:11]: 👍🏼

File: test/tests/parallel/set1/spam_virus/fail2ban.bats
Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/fail2ban.bats:25]: No, that was by accident.

File: test/tests/parallel/set1/spam_virus/fail2ban.bats
Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/fail2ban.bats:200]: Yes, please; let's do this in another PR :)

File: test/tests/parallel/set1/spam_virus/postgrey_enabled.bats
Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/postgrey_enabled.bats:21]: I will add it back, you're probably right about this.

File: test/tests/parallel/set1/spam_virus/spam_junk_folder.bats
Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/spam_junk_folder.bats:12]: What exactly should? Probably this is related to <https://github.com/docker-mailserver/docker-mailserver/pull/2938/files#r1056041092> and hence, somewhat resolved?

**Comment by georglauterbach** [comment]
@polarathene I resolved most of the your feedback and added your proposed changes. Then I resolved the conflict by rebasing on the current `master`. Finally, added one more helper aboput whether a process is running or not and adjusted the SSL test according to your feedback.

Open tasks for you are now:

1. Double-check the SSL test(s) (in case they are failing, this is obvious, but if they succeed now - and they do locally for me - check whether all _really_ everything is as expected)
2. Rewrite history according to your liking :)

File: test/tests/parallel/set1/spam_virus/fail2ban.bats
Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/fail2ban.bats]: Done in 0902ab2f4ec2bb73350039702ba7b6cd111a9ac3

File: test/tests/parallel/set1/spam_virus/spam_junk_folder.bats
Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/spam_junk_folder.bats]: Done in cfc4a5aec8563a95e3fbc6069f81f9036acb4d4f

File: test/tests/parallel/set1/spam_virus/fail2ban.bats
Comment (georglauterbach) [reply test/tests/parallel/set1/spam_virus/fail2ban.bats]: Done in 0902ab2f4ec2bb73350039702ba7b6cd111a9ac3

File: test/tests/parallel/set1/spam_virus/clamav.bats
Comment (polarathene) [reply test/tests/parallel/set1/spam_virus/clamav.bats:29]: I won't block on this as it's not relevant to the test, although for that same reason since I think that's the only change in that file causing all the diffs, it could be reverted too :man_shrugging: 

After we've converted all the serial tests to leverage the `setup.bash` helper, we can be nitpicky about consistency and test case names I guess :sweat_smile: 

Resolving as non-blocking concern.

File: test/tests/parallel/set1/spam_virus/disabled_clamav_spamassassin.bats
Comment (polarathene) [reply test/tests/parallel/set1/spam_virus/disabled_clamav_spamassassin.bats]: This is resolved and I agree with deferring. Just keeping it "unresolved" for visibility of something that should be addressed in future.

File: test/tests/parallel/set1/spam_virus/spam_junk_folder.bats
Comment (polarathene) [reply test/tests/parallel/set1/spam_virus/spam_junk_folder.bats:14]: > What exactly should?

`CONTAINER_NAME` in most of your converted tests is defined statically (_or variants are `CONTAINER_NAME1`, `CONTAINER_NAME2`, and re-assigned to `CONTAINER_NAME` in a test case_) at the top right after `TEST_NAME_PREFIX` is defined.

---

I was pointing out that in this file you seemed to have been inconsistent with that convention. The referenced feedback you asked about is hand-waving the concern about `CONTAINER_NAME` usage for assertions instead of `CONTAINER_NAME1` or similar as all the assertions are within the same test case definition, so I'm not as fussed as I was when I initially read them thinking there was more than 2 test cases present.

I am not blocking on this, just an observation that appears to be breaking convention for convenience.

File: test/tests/parallel/set1/spam_virus/spam_junk_folder.bats
Comment (polarathene) [reply test/tests/parallel/set1/spam_virus/spam_junk_folder.bats:41]: Yes, not a problem in current usage :+1:

File: test/tests/parallel/set1/spam_virus/fail2ban.bats
Comment (polarathene) [reply test/tests/parallel/set1/spam_virus/fail2ban.bats]: > Uhhh, ohhh; I just wrote the above text, searched for it and found the solution: `--no-parallelize-within-files`. I will provide an update for this.

I haven't looked at the related docs in some time, but I think there may have been a way to selectively indicate to BATS which files should opt-out of this kind of parallelization.

For now though without looking through each files test cases in sequence, it's probably better to opt-out completely like you've gone ahead and done :+1: 

I know with a test like the cipherlists where each test case takes a fair chunk of time on it's own (_in total I think it's about 25-33% of the serial test time due to some notable slowdown overhead caused by dockers networking_), it'd be useful to run each test case in parallel. But with your efforts here, it looks like we could also split out common logic into a helper and have each test case as it's own file grouped in a folder ;)

---

Feel free to resolve this once you've read the message :+1:

**Comment by polarathene** [comment]
> and adjusted the SSL test according to your feedback.

This was not necessary as I had already taken care of it sorry :sweat_smile: (_I either didn't communicate that well, or relied on the added commit history to do that itself_)

> This PR roughly shoves off 7 minutes from the serial tests: takes 23 now, was 30 before :)

Fantastic news! We can probably cover the cipherlists test here as well to bring that down closer to 15 for serial. I'll handle that :+1: 

---

I'll start rewriting the history now :)

File: docs/content/contributing/general.md
Comment (georglauterbach) [reply docs/content/contributing/general.md:31]: Ran some tests on my system; I have an older i7-4770k with 8 threads / 4 cores. Running with `--jobs 4` was the maximum when it came to CPU utilisation for me. I'd rather set the bar high and avoid unnecessary test failures due to resource problems than setting it low and risking them. I used `htop` to look at CPU/RAM.

File: Makefile
Comment (georglauterbach) [reply Makefile:59]: You're probably right; using just `BATS_FLAGS` instead of `BATS_FLAGS_PARALLEL` is better in this case. You can just update this as well :)

File: docs/content/contributing/general.md
Comment (polarathene) [reply docs/content/contributing/general.md:31]: I used `--jobs 4` (_with and without `--no-parallelize-within-files`_) on my i5-6500 system (_32GB RAM, 16GB in use_) and for `parallel/set1` I only observed near 2GB peak usage increase via system monitor and reaching 100% CPU load.

The only reason I think you'd have experienced higher RAM usage would perhaps be if the image under test hadn't been built with a recent ClamAV layer updated. I have run tests with the last clean image build being a week or so old and observed ClamAV tests were very memory hungry.

Might have been the case for you? That or there's something different with the latest changes I've pushed for this branch :man_shrugging: 

---

So I'm still not sure how high of a bar you want to set that. Before parallel tests were added, I've seen tests with ClamAV under some circumstances gobble up 6GB+ in memory :man_shrugging: 

Do you have a source for each increment of the `BATS_PARALLEL_JOBS` / `--jobs` requiring double the number of threads? It doesn't seem to be mentioned on the [BATS docs](https://bats-core.readthedocs.io/en/v1.8.2/usage.html).

My understanding is it's just a concurrent queue of jobs that'll get scheduled CPU time on available threads, so since N jobs doesn't necessarily use constant CPU, more CPU time can be available for additional jobs to run?

**Comment by polarathene** [comment]
Apologies for the delay.

- I'm happy to rebase squash the docs commit into `tests: Adjusted files not directly related to tests`. I've just kept it separate for review feedback as I revised it a bit. [Preview page](https://pullrequest-2938--dms-doc-previews.netlify.app/contributing/general/). Still discussing about a portion of this content in earlier review comment [here](https://github.com/docker-mailserver/docker-mailserver/pull/2938#discussion_r1060358626).
- **I've not only rewritten history**, other changes have been added and compressed into the commits. The first is minimal, a 2nd commit applies larger refactoring and could be optionally extracted out of this PR as a follow-up. Each commit provides details on relevant changes.
- There was a notable gotcha with the `_run_in_container` method @georglauterbach added in this PR that I've included a comment next to it's declaration to keep in mind. I also needed to update `check_if_process_is_running` to not use that method as it was leading to tests passing (_restart when killed test cases_) when they actually were failing :sweat_smile:

File: docs/content/contributing/general.md
Comment (georglauterbach) [reply docs/content/contributing/general.md:31]: Seems as if the resource requirements differ a lot. The ClamAV update could be an issue, but I don't think so since I built it directly before running the tests, IIRC.

I have no source when it comes to the `--jobs` flag needing more threads. We should therefore probably refrain from using exact numbers here and be more abstract. Something like: "Depending on the hardware in your system, you can increase the number of parallel jobs. Make sure you do not overload the system, which can lead to failing tests."

**Review by casperklein — APPROVED** [review]

**Review by polarathene — APPROVED** [review]

**Comment by polarathene** [comment]
Thanks @georglauterbach for tackling this and @casperklein for taking the time to review :tada:

I'll follow-up with a few more similar PRs :muscle:

**Pull request merged**
